security/advancedtls/advancedtls.go

```diff
@@ -187,39 +205,105 @@ func (o *ClientOptions) config() (*tls.Config, error) {
 		return nil, fmt.Errorf(
 			"client needs to provide custom verification mechanism if choose to skip default verification")
 	}
+	// Make sure users didn't specify more than one fields in
+	// RootCertificateOptions and IdentityCertificateOptions.
+	rootOptionNum := o.RootCertificateOptions.nonNilFieldCount()
+	if rootOptionNum > 1 {
+		return nil, fmt.Errorf(
+			"at most one field in RootCertificateOptions could be specified")
+	}
+	identityOptionNum := o.IdentityCertificateOptions.nonNilFieldCount()
+	if identityOptionNum > 1 {
+		return nil, fmt.Errorf(
+			"at most one field in IdentityCertificateOptions could be specified")
+	}
+	if o.GetIdentityCertificatesForServer != nil {
+		return nil, fmt.Errorf(
+			"GetIdentityCertificatesForServer cannot be specified on the client side")
+	}
+	// Use system default trust certificate if no verification specified.
 	rootCAs := o.RootCACerts
-	if o.VType != SkipVerification && o.RootCACerts == nil && o.GetRootCAs == nil {
+	if o.VType != SkipVerification && rootOptionNum == 0 {
 		// Set rootCAs to system default.
 		systemRootCAs, err := x509.SystemCertPool()
 		if err != nil {
 			return nil, err
 		}
 		rootCAs = systemRootCAs
 	}
-	// We have to set InsecureSkipVerify to true to skip the default checks and
-	// use the verification function we built from buildVerifyFunc.
+	// Fill the root and identity reloading functions based on the Provider
+	// implementation.
+	if o.RootProvider != nil {
+		o.GetRootCertificates = func(*GetRootCAsParams) (*GetRootCAsResults, error) {
+			ctx, cancel := context.WithCancel(context.Background())
```

Comment:
This can block forever if the `Provider` does not provide any key material.

Reply:
Can I know the possible ways to solve this? Maybe add a deadline to this request?
I think from the API's perspective, if users use a Provider without feeding in any contents, that might be a misuse of the API. Adding a deadline might mitigate the issue. I'd just like to know if there are any other alternatives.

Reply:
I think it might be ok to not specify a deadline here. But this should be made obvious to the user. Because eventually, their `grpc.Dial()` will exceed its deadline, and they will call `Close` on their `Provider` which will unblock this call.

If you want to support adding a deadline here, that needs to be plumbed through the API. Maybe something for you to consider.

Reply:
SG. I think it's a requirement in our case to have initial bootstrapping credentials before we start connection, so no need to add  a deadline here. I will document in the comments saying explicitly that Provider needs to have some contents.